harvester_e2e_tests/integration/test_vm.py

```diff
@@ -0,0 +1,295 @@
+    )
+    code, data = api_client.vms.migrate(second_vm_name, second_vm_dst_host)
+    assert code == 204, (
+        f"Can't migrate VM {second_vm_name} from host {second_vm_src_host} to {second_vm_dst_host}"
+    )
+
+    endtime = datetime.now() + timedelta(seconds=wait_timeout)
+    while endtime > datetime.now():
+        code, data = api_client.vms.get_status(first_vm_name)
+        if data.get('status', {}).get('migrationState', {}).get('completed', False):
+            assert first_vm_dst_host == data['status']['nodeName'], (
+                f"Failed to migrate VM {first_vm_name} \
+                    from {first_vm_src_host} to {first_vm_dst_host}"
+            )
+            break
+        sleep(5)
+    else:
+        raise AssertionError(
+            f"The migration of VM {first_vm_name} is not completed with {wait_timeout} timed out"
+            f"Got error: {code}, {data}"
+        )
+    endtime = datetime.now() + timedelta(seconds=wait_timeout)
+    while endtime > datetime.now():
+        code, data = api_client.vms.get_status(second_vm_name)
+        if data.get('status', {}).get('migrationState', {}).get('completed', False):
+            assert second_vm_dst_host == data['status']['nodeName'], (
+                f"Failed to migrate VM {second_vm_name} \
+                    from {second_vm_src_host} to {second_vm_dst_host}"
+            )
+            break
+        sleep(5)
+    else:
+        raise AssertionError(
+            f"The migration of VM {second_vm_name} is not completed with {wait_timeout} timed out"
+            f"Got error: {code}, {data}"
+        )
+
+    # teardown
+    api_client.vms.delete(first_vm_name)
+    volume_name = ""
+    for volume in first_vm_data['spec']['volumes']:
+        if volume['name'] == 'disk-0':
+            volume_name = volume['persistentVolumeClaim']['claimName']
+    api_client.volumes.delete(volume_name)
+    api_client.vms.delete(second_vm_name)
+    for volume in second_vm_data['spec']['volumes']:
+        if volume['name'] == 'disk-0':
+            volume_name = volume['persistentVolumeClaim']['claimName']
+    api_client.volumes.delete(volume_name)
+
+
+@pytest.mark.singlevmtest
+@pytest.mark.virtual_machines_p1
+def test_migrate_vm_with_user_data(api_client, unique_name, focal_image, wait_timeout,
+                                   available_node_names):
+    vm_spec = api_client.vms.Spec(1, 1)
+    vm_spec.add_image('disk-0', focal_image['id'])
+    user_data = vm_spec.user_data
+    user_data += (
+        "password: test\n"
+        "chpasswd:\n"
+        "  expire: false\n"
+        "ssh_pwauth: true\n"
+    )
+    vm_spec.user_data = user_data
+    code, data = api_client.vms.create(unique_name, vm_spec)
+    assert code == 201, (
+        f"Failed to create VM {unique_name} with error: {code}, {data}"
+    )
+
+    wait_vm_agent_connected(api_client, unique_name, wait_timeout)
```

Comment:
If we really need helper function, we should use assertion in test cases, not in helper functions.